{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#exports\n",
    "\n",
    "import pysam\n",
    "from Bio.Seq import Seq\n",
    "from script.utils import Vividict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exports\n",
    "\n",
    "def split_bed_line(entry, extends=False):\n",
    "\t\"\"\"\n",
    "\tsplit bed line\n",
    "\t\"\"\"\n",
    "\tcells = entry.strip('\\n').split(\"\\t\")\n",
    "\tchrom = start = end = name = score = strand = thick_start = thick_end = item_rgb = block_count = block_sizes = block_starts = None\n",
    "\tif len(cells) >= 6:\n",
    "\t\tchrom = cells[0]\n",
    "\t\tstart = int(cells[1])\n",
    "\t\tend = int(cells[2])\n",
    "\t\tname = cells[3]\n",
    "\t\tscore = cells[4]\n",
    "\t\tstrand= cells[5]\n",
    "\tif len(cells) >=12:\n",
    "\t\tthick_start = int(cells[6])\n",
    "\t\tthick_end = int(cells[7])\n",
    "\t\titem_rgb = cells[8]\n",
    "\t\tblock_count = int(cells[9])\n",
    "\t\tblock_sizes = [int(i) for i in cells[10].rstrip(',').split(',')]\n",
    "\t\tblock_starts= [int(i) for i in cells[11].rstrip(',').split(',')]\n",
    "\t\t\n",
    "\treturn chrom,start,end,name,score,strand,thick_start,thick_end,item_rgb,block_count,block_sizes,block_starts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exports\n",
    "\n",
    "def bed_block_to_splicing(start, block_count, block_starts, block_sizes):\n",
    "\t\"\"\"convert bed block to isoform splicing structure\n",
    "\t\"\"\"\n",
    "\tread_splicing = []\n",
    "\tif block_count > 1:\n",
    "\t\tfor i in range(block_count-1):\n",
    "\t\t\tleft_sj = start + block_starts[i] + block_sizes[i]\n",
    "\t\t\tright_sj = start + block_starts[i+1] + 1\n",
    "\t\t\tread_splicing.extend([left_sj, right_sj])\n",
    "\treturn read_splicing\n",
    "\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exports\n",
    "\n",
    "def read_grouping(bed_file, fasta):\n",
    "\t\"\"\"\n",
    "\treturn junctions with only one long read support\n",
    "\t\"\"\"\n",
    "\tjunction_dict = Vividict()\n",
    "\tprocessed_read = Vividict()\n",
    "\tgenome = pysam.FastaFile(fasta)\n",
    "\twith open (bed_file) as f:\n",
    "\t\tfor read in f:\n",
    "\t\t\tchrom, start, end, name, score, strand, thick_start, thick_end, item_rgb, block_count, block_sizes, block_starts = split_bed_line(read) # read the each nanopore read\n",
    "\t\t\tfull_block = [start+1]\n",
    "\t\t\tif block_count > 1:\n",
    "\t\t\t\tfor i in range(block_count-1):\n",
    "\t\t\t\t\tleft_sj = start + block_starts[i] + block_sizes[i]\n",
    "\t\t\t\t\tright_sj = start + block_starts[i+1] + 1\n",
    "\t\t\t\t\tfull_block.extend([left_sj, right_sj])\n",
    "\t\t\t\t\ttmp_sj = (chrom,strand,left_sj,right_sj)\n",
    "\t\t\t\t\t\n",
    "\t\t\t\t\tif tmp_sj not in junction_dict[(chrom,strand)]:\n",
    "\t\t\t\t\t\tif strand =='+':\n",
    "\t\t\t\t\t\t\tintron_motif = f'{genome.fetch(chrom, left_sj, left_sj+2)}-{genome.fetch(chrom, right_sj-3, right_sj-1)}'\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\tintron_motif = f'{Seq(genome.fetch(chrom, right_sj-3, right_sj-1)).reverse_complement()}-{Seq(genome.fetch(chrom, left_sj, left_sj+2)).reverse_complement()}'\n",
    "\t\t\t\t\t\tif intron_motif in ['GT-AG', 'CT-AC', 'GC-AG', 'CT-GC', 'AT-AC', 'GT-AT']:\n",
    "\t\t\t\t\t\t\tjunction_dict[(chrom,strand)][tmp_sj] = [1, intron_motif, 'canonical']\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\tjunction_dict[(chrom,strand)][tmp_sj] = [1, intron_motif, 'non-canonical']\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tjunction_dict[(chrom,strand)][tmp_sj][0] += 1\n",
    "\t\t\tfull_block.append(end)\n",
    "\t\t\tname = f'{start+1}_{name}'\n",
    "\t\t\tprocessed_read[(chrom,strand)][name] = full_block\n",
    "\t\t\t\n",
    "\treturn junction_dict, processed_read\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058583.sorted.bed'\n",
    "fa = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta_170612a.fasta'\n",
    "junction_dict, processed_read = read_grouping(bed, fa)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exports\n",
    "\n",
    "def polya_signal_import(polyadenylation_event):\n",
    "\t\"\"\"\n",
    "\tread the polyA information from nanopolish-polya mode\n",
    "\t\"\"\"\n",
    "\tpolya_reads = {}\n",
    "\tif polyadenylation_event:\n",
    "\t\twith open (polyadenylation_event) as f:\n",
    "\t\t\tnext(f)\n",
    "\t\t\tfor line in f:\n",
    "\t\t\t\tline = line.rstrip().split(\"\\t\")\n",
    "\t\t\t\tif len(line) != 10:\n",
    "\t\t\t\t\traise ValueError(\"Fatal: input polyA results was not from nanopolish!\")\n",
    "\t\t\t\telif line[9] == 'PASS': \n",
    "\t\t\t\t\tpolya_reads[f'{int(line[2])+1}_{line[0]}'] = True\n",
    "\treturn polya_reads\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exports\n",
    "\n",
    "class PolyAFinder():\n",
    "\t\"\"\"estimate read polyadenylation event by checking the polya motifs from sequence\n",
    "\t\"\"\"\n",
    "\tdef __init__(self, processed_read, fasta, polyA_motif_file, updis = 40,  downdis = 10):\n",
    "\t\tself.processed_read = processed_read\n",
    "\t\tself.fasta = fasta\n",
    "\t\tself.polyA_motif_file = polyA_motif_file\n",
    "\t\tself.updis = updis\n",
    "\t\tself.downdis = downdis\n",
    "\n",
    "\tdef polyA_motif_import(self):\n",
    "\t\t\"\"\"import given polyA motif\n",
    "\t\t\"\"\"\n",
    "\t\tpolyA_motif_list = []\n",
    "\t\tfor line in open(self.polyA_motif_file):\n",
    "\t\t\tx = line.strip().upper().replace('U', 'A')\n",
    "\t\t\tpolyA_motif_list.append(x)\n",
    "\n",
    "\t\treturn polyA_motif_list\n",
    "\n",
    "\tdef find_polyA_motif(self, seq, polyA_motif_list):\n",
    "\t\t\"\"\"check the occurrence of given polyA motif in target sequence\n",
    "\t\t\"\"\"\n",
    "\t\tfor motif in polyA_motif_list:\n",
    "\t\t\ti = seq.find(motif)\n",
    "\t\t\tif i >= 0:\n",
    "\t\t\t\treturn motif\n",
    "\t\treturn None\n",
    "\n",
    "\tdef polya_estimation(self):\n",
    "\t\t\"\"\"estimate read polyadenylation event\n",
    "\t\t\"\"\"\n",
    "\t\tgenome = pysam.FastaFile(self.fasta)\n",
    "\t\tread_polya_dict = {}\n",
    "\t\tpolyA_motif_list = self.polyA_motif_import()\n",
    "\n",
    "\t\tfor (chrom, strand) in self.processed_read:\n",
    "\t\t\tfor read_id, full_block in self.processed_read[(chrom, strand)].items():\n",
    "\n",
    "\t\t\t\tif strand == '-':\n",
    "\t\t\t\t\tres = full_block[0]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tstrand = '+'\n",
    "\t\t\t\t\tres = full_block[-1]\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tif strand == '-':\n",
    "\t\t\t\t\t\tseq = f'{Seq(genome.fetch(chrom, res - self.downdis, res + self.updis)).reverse_complement()}'\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tseq = f'{genome.fetch(chrom, res - self.updis, res + self.downdis)}'\n",
    "\t\t\t\t\tpolyA_motif = self.find_polyA_motif(seq, polyA_motif_list)\n",
    "\t\t\n",
    "\t\t\t\t\tif polyA_motif:\n",
    "\t\t\t\t\t\tread_polya_dict[read_id] = True\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\tpass\n",
    "\t\treturn read_polya_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
