{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "output-file: wrapper.html\n",
    "title: lafite_wrapper\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp lafite_wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "import sys ,os, subprocess\n",
    "import argparse\n",
    "\n",
    "from numpy import percentile\n",
    "from time import strftime\n",
    "\n",
    "from LAFITE.reference_processing import RefProcessWrapper, short_reads_sj_import, cage_tss_import\n",
    "from LAFITE.preprocessing import read_grouping, polya_signal_import, PolyAFinder\n",
    "from LAFITE.utils import temp_dir_creation, bam2bed, keep_tmp_file\n",
    "from LAFITE.read_collapsing import CoCoWrapper\n",
    "from LAFITE.tailFinder import TailFinderWrapper\n",
    "from LAFITE.refine import RefineWrapper\n",
    "from LAFITE.output import OutputAssembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "class LafiteWrapper:\n",
    "\tdef __init__(self, bam, bedtools, full_cleanup, gtf, genome, min_count_tss_tes, mis_intron_length, min_novel_trans_count, min_single_exon_coverage, min_single_exon_len, label, output, polya, polyA_motif_file, relative_abundance_threshold, sj_correction_window, short_sj_tab, thread, tss_cutoff, tss_peak):\n",
    "\t\tself.bam = bam\n",
    "\t\tself.bedtools = bedtools\n",
    "\t\tself.full_cleanup = full_cleanup\n",
    "\t\tself.gtf = gtf\n",
    "\t\tself.genome = genome\n",
    "\t\tself.min_count_tss_tes = min_count_tss_tes\n",
    "\t\tself.mis_intron_length = mis_intron_length\n",
    "\t\tself.min_novel_trans_count = min_novel_trans_count\n",
    "\t\tself.min_single_exon_coverage = min_single_exon_coverage\n",
    "\t\tself.min_single_exon_len = min_single_exon_len\n",
    "\t\tself.label = label\n",
    "\t\tself.output = output\n",
    "\t\tself.polya = polya\n",
    "\t\tself.polyA_motif_file = polyA_motif_file\n",
    "\t\tself.relative_abundance_threshold = relative_abundance_threshold\n",
    "\t\tself.sj_correction_window = sj_correction_window\n",
    "\t\tself.short_sj_tab = short_sj_tab\n",
    "\t\tself.thread = thread\n",
    "\t\tself.tss_cutoff = tss_cutoff\n",
    "\t\tself.tss_peak = tss_peak\n",
    "\t\t\n",
    "\t\n",
    "\tdef revisit_parameter(self):\n",
    "\t\t\"\"\"\n",
    "\t\trevisit input parameters\"\"\"\n",
    "\n",
    "\t\tsys.stdout.write(f'\\nInput parameters:\\n')\n",
    "\t\tsys.stdout.write(f'Read alignment file: {self.bam}\\n')\n",
    "\t\tsys.stdout.write(f'Reference gene annotation: {self.gtf}\\n')\n",
    "\t\tsys.stdout.write(f'Reference genome annotation: {self.genome}\\n')\n",
    "\t\tif self.polya:\n",
    "\t\t\tsys.stdout.write(f'Reads Polyadenylation events: {self.polya}\\n')\n",
    "\t\telif self.polyA_motif_file:\n",
    "\t\t\tsys.stdout.write(f'PolyA motif file for read Polyadenylation event estimation: {self.polyA_motif_file}\\n')\n",
    "\t\telse:\n",
    "\t\t\traise ValueError('Fatal: Please provide either polyA motif file or reads polyadenylation event result\\n')\n",
    "\t\tsys.stdout.write(f'Edit distance to reference splicing site allowed for splicing correction: {self.sj_correction_window}\\n')\n",
    "\t\tsys.stdout.write(f'Output assembly file: {self.output}\\n')\n",
    "\n",
    "\tdef run_lafite(self):\n",
    "\t\t\"\"\"\n",
    "\t\tLAFITE wrapper\"\"\"\n",
    "\n",
    "\t\tself.revisit_parameter()\n",
    "\t\t# create temp directory for log and intermediate files\n",
    "\t\ttry:\n",
    "\t\t\ttmp_folder = temp_dir_creation(os.path.dirname(self.output))\n",
    "\t\t\ttmp_dir = tmp_folder.name\n",
    "\t\texcept:\n",
    "\t\t\traise ValueError('Fatal: Please provide a valid path for output files\\n')\n",
    "\n",
    "\t\t# reference gene annotation processing\n",
    "\t\tsys.stdout.write(strftime(\"%Y-%m-%d %H:%M:%S\") + ': Preprocessing reference gene annotation\\n')\n",
    "\t\tref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = RefProcessWrapper(self.gtf, self.thread).result_collection()\n",
    "\n",
    "\t\tif self.short_sj_tab:\n",
    "\t\t\tleft_sj_set, right_sj_set = short_reads_sj_import(self.short_sj_tab, left_sj_set, right_sj_set)\n",
    "\t\t\n",
    "\t\tif self.tss_peak:\n",
    "\t\t\ttss_dict = cage_tss_import(self.tss_peak, tss_dict)\n",
    "\n",
    "\t\t# processing alignment bam file\n",
    "\t\t## convert alignment bam file to bed12 format\n",
    "\t\tsys.stdout.write(strftime(\"%Y-%m-%d %H:%M:%S\") + ': Preprocessing alignment file\\n')\n",
    "\t\ttry: \n",
    "\t\t\tbam2bed_cmd = bam2bed(self.bam, tmp_dir, self.bedtools)\n",
    "\t\texcept:\n",
    "\t\t\traise ValueError('Fatal: Please provide a valid path for bam file and bedtools\\n')\n",
    "\t\tp = subprocess.run(bam2bed_cmd, shell=True)\n",
    "\t\tif p.returncode == 0:\n",
    "\t\t\tpass\n",
    "\t\telse:\n",
    "\t\t\traise ValueError('Fatal: Error in bam conversion\\n')\n",
    "\n",
    "\t\t## read grouping according to chromosome and strand\n",
    "\t\toutbed = os.path.join(tmp_dir, 'bam.bed')\n",
    "\t\tjunction_dict, processed_read = read_grouping(outbed, self.genome)\n",
    "\n",
    "\t\t# polyA info import\n",
    "\t\tif self.polya:\n",
    "\t\t\tsys.stdout.write(strftime(\"%Y-%m-%d %H:%M:%S\") + ': Loading polyA information\\n')\n",
    "\t\t\tpolya_dict = polya_signal_import(self.polya)\n",
    "\t\telse:\n",
    "\t\t\tsys.stdout.write(strftime(\"%Y-%m-%d %H:%M:%S\") + ': No reads Polyadenylation event provided, detecting from sequence\\n')\n",
    "\t\t\tpolya_dict = PolyAFinder(processed_read, self.genome, self.polyA_motif_file).polya_estimation()\n",
    "\n",
    "\t\t# read correction and collapsing\n",
    "\t\tsys.stdout.write(strftime(\"%Y-%m-%d %H:%M:%S\") + ': Collapssing corrected reads\\n')\n",
    "\t\tcollected_single_exon_read, collected_multi_exon_read, collected_rss, collected_res = CoCoWrapper(self.thread, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, self.sj_correction_window, polya_dict, self.mis_intron_length, tmp_dir).result_collection()\n",
    "\n",
    "\t\t# identify putative TSS and TES for collapsed reads\n",
    "\t\tprocessed_collected_multi_exon_read, three_prime_exon = TailFinderWrapper(collected_multi_exon_read, self.min_count_tss_tes, self.thread).result_collection()\n",
    "\n",
    "\t\t# calculating the tss_cutoff and tes_cutoff:\n",
    "\t\tif not self.tss_cutoff:\n",
    "\t\t\tself.tss_cutoff = percentile(collected_rss, 75)\n",
    "\t\ttes_cutoff = percentile(collected_res, 70)\n",
    "\t\tprint(f'TSS cutoff: {self.tss_cutoff}')\n",
    "\t\tprint(f'TES cutoff: {tes_cutoff}')\n",
    "\n",
    "\t\t# identify high quality isoforms from collapsed reads\n",
    "\t\tsys.stdout.write(strftime(\"%Y-%m-%d %H:%M:%S\") + ': Revisiting the collapsed reads to get high-concensus full-length isoforms\\n')\n",
    "\t\tcollected_refined_isoforms = RefineWrapper(processed_collected_multi_exon_read, collected_single_exon_read, ref_mutple_exon_trans, ref_single_exon_trans, three_prime_exon, tss_dict, self.tss_cutoff, tes_cutoff, self.min_novel_trans_count, self.min_single_exon_coverage, self.min_single_exon_len, self.thread, tmp_dir).result_collection()\n",
    "\n",
    "\t\t# output refined isoforms\n",
    "\t\tOutputAssembly(collected_refined_isoforms, self.output, self.label, self.relative_abundance_threshold).write_out()\n",
    "\n",
    "\t\t# keep intermediate results\n",
    "\t\tif self.full_cleanup:\n",
    "\t\t\tos.system(keep_tmp_file(self.output, tmp_dir))\n",
    "\t\t\n",
    "\t\tsys.stdout.write(strftime(\"%Y-%m-%d %H:%M:%S\") + ': All done!\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "def main():\n",
    "\n",
    "\tparser = argparse.ArgumentParser(description='Low-abundance Aware Full-length Isoform clusTEr')\n",
    "\tparser.add_argument('-b', dest='bam', help='path to the alignment file in bam format', required=True)\n",
    "\tparser.add_argument('-B', dest='bedtools', type=str, default='bedtools', help='path to the executable bedtools')\n",
    "\tparser.add_argument('-g', dest='gtf', help='path to the reference gene annotation in GTF format', required=True)\n",
    "\tparser.add_argument('-f', dest='genome', help='path to the reference genome fasta', required=True)\n",
    "\tparser.add_argument('-o', dest='output', help='path to the output file', required=True)\n",
    "\tparser.add_argument('-n', dest='min_count_tss_tes', type=int, default=3, help='minimum number of reads supporting a alternative TSS or TES, default: 3')\n",
    "\tparser.add_argument('-i', dest='mis_intron_length', type=int, default=150, help='length cutoff for correcting unexpected small intron, default: 150')\n",
    "\tparser.add_argument('-c', dest='min_novel_trans_count', type=int, default=3, help='minimum occurrences required for a isoform from novel loci, default: 3')\n",
    "\tparser.add_argument('-s', dest='min_single_exon_coverage', type=int, default=4, help='minimum read coverage required for a novel single-exon transcript, default: 4')\n",
    "\tparser.add_argument('-l', dest='min_single_exon_len', type=int, default=100, help='minimum length for single-exon transcript, default: 100')\n",
    "\tparser.add_argument('-L', dest='label', type=str, default='LAFT', help='name prefix for output transcripts, default: LAFT')\n",
    "\tparser.add_argument('-p', dest='polya', help='path to the file contains read Polyadenylation event')\n",
    "\tparser.add_argument('-m', dest='polyA_motif_file', help='path to the polya motif file')\n",
    "\tparser.add_argument('-r', dest='relative_abundance_threshold', type=int, default=0.01, help='minimum abundance of the predicted multi-exon transcripts as a fraction of the total transcript assembled at a given locus, default: 0.01')\n",
    "\tparser.add_argument('-j', dest='short_sj_tab', default=None, help='path to the short read splice junction file')\n",
    "\tparser.add_argument('-w', dest='sj_correction_window', type=int, default=40, help='edit distance to reference splicing site for splicing correction, default: 40')\n",
    "\tparser.add_argument('--no_full_cleanup', dest='full_cleanup', action='store_true', help='keep all intermediate files')\n",
    "\tparser.add_argument('-t', dest='thread', type=int, default=4, help='number of the threads, default: 4')\n",
    "\tparser.add_argument('-T', dest='tss_peak', default=None, help='path to the TSS peak file')\n",
    "\tparser.add_argument('-d', dest='tss_cutoff', type=int, default=None, help='minimum TSS distance for a transcript to be considered as a novel transcript')\n",
    "\targs = parser.parse_args()\n",
    "\n",
    "\tLafiteWrapper(args.bam, args.bedtools, args.full_cleanup, args.gtf, args.genome, args.min_count_tss_tes, args.mis_intron_length, args.min_novel_trans_count, args.min_single_exon_coverage, args.min_single_exon_len, args.label, args.output, args.polya, args.polyA_motif_file, args.relative_abundance_threshold, args.sj_correction_window, args.short_sj_tab, args.thread, args.tss_cutoff, args.tss_peak).run_lafite()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gtf = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf'\n",
    "bed = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058583.sorted.bed'\n",
    "fa = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta_170612a.fasta'\n",
    "junction_dict, processed_read = read_grouping(bed, fa)\n",
    "ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = RefAnnotationExtraction(gtf).annotation_sorting()\n",
    "\n",
    "# polya_dict = polya_signal_import('/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058584.polya.res')\n",
    "# polya_dict = PolyAFinder(processed_read, fa, '/home/zjzace/software/SQANTI3-4.1/data/polyA_motifs/mouse_and_human.polyA_motif.txt').polya_estimation()\n",
    "# collected_single_exon_read, collected_multi_exon_read, tss, rss = CoCoWrapper(16, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, tmp_dir='.', sj_correction_window=40, mis_intron_length = 150, corExcept_dis=4, polya_dict=polya_dict).result_collection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "!nbdev_export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm 'Makefile'\n"
     ]
    }
   ],
   "source": [
    "!git rm Makefile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "find . -name '*.ipynb' -exec perl -pi -e 's/#\\|\\s*hide_input/#| echo: false/' {} +\n",
    "find . -name '*.ipynb' -exec perl -pi -e 's/#\\|\\s*hide_output/#| output: false/' {} +\n",
    "find . -name '*.ipynb' -exec perl -pi -e 's/#\\|\\s*skip/#| eval: false/' {} +\n",
    "find . -name '*.ipynb' -exec perl -pi -e 's/from nbdev.export import notebook2script/from nbdev import nbdev_export/' {} +\n",
    "find . -name '*.ipynb' -exec perl -pi -e 's/notebook2script/nbdev_export/' {} +"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gtf = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf'\n",
    "bed = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058583.sorted.bed'\n",
    "fa = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta_170612a.fasta'\n",
    "junction_dict, processed_read = read_grouping(bed, fa)\n",
    "ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = RefAnnotationExtraction(gtf).annotation_sorting()\n",
    "\n",
    "# polya_dict = polya_signal_import('/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058584.polya.res')\n",
    "# polya_dict = PolyAFinder(processed_read, fa, '/home/zjzace/software/SQANTI3-4.1/data/polyA_motifs/mouse_and_human.polyA_motif.txt').polya_estimation()\n",
    "# collected_single_exon_read, collected_multi_exon_read, tss, rss = CoCoWrapper(16, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, tmp_dir='.', sj_correction_window=40, mis_intron_length = 150, corExcept_dis=4, polya_dict=polya_dict).result_collection()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
