---

title: lafite_wrapper


keywords: fastai
sidebar: home_sidebar



nb_path: "10_wrapper.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 10_wrapper.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span> <span class="o">,</span><span class="nn">os</span><span class="o">,</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">argparse</span>

<span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">percentile</span>
<span class="kn">from</span> <span class="nn">time</span> <span class="kn">import</span> <span class="n">strftime</span>

<span class="kn">from</span> <span class="nn">LAFITE.reference_processing</span> <span class="kn">import</span> <span class="n">RefProcessWrapper</span><span class="p">,</span> <span class="n">short_reads_sj_import</span><span class="p">,</span> <span class="n">cage_tss_import</span>
<span class="kn">from</span> <span class="nn">LAFITE.preprocessing</span> <span class="kn">import</span> <span class="n">read_grouping</span><span class="p">,</span> <span class="n">polya_signal_import</span><span class="p">,</span> <span class="n">PolyAFinder</span>
<span class="kn">from</span> <span class="nn">LAFITE.utils</span> <span class="kn">import</span> <span class="n">temp_dir_creation</span><span class="p">,</span> <span class="n">bam2bed</span><span class="p">,</span> <span class="n">keep_tmp_file</span>
<span class="kn">from</span> <span class="nn">LAFITE.read_collapsing</span> <span class="kn">import</span> <span class="n">CoCoWrapper</span>
<span class="kn">from</span> <span class="nn">LAFITE.tailFinder</span> <span class="kn">import</span> <span class="n">TailFinderWrapper</span>
<span class="kn">from</span> <span class="nn">LAFITE.refine</span> <span class="kn">import</span> <span class="n">RefineWrapper</span>
<span class="kn">from</span> <span class="nn">LAFITE.output</span> <span class="kn">import</span> <span class="n">OutputAssembly</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gtf</span> <span class="o">=</span> <span class="s1">&#39;/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf&#39;</span>
<span class="n">bed</span> <span class="o">=</span> <span class="s1">&#39;/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058583.sorted.bed&#39;</span>
<span class="n">fa</span> <span class="o">=</span> <span class="s1">&#39;/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta_170612a.fasta&#39;</span>
<span class="n">junction_dict</span><span class="p">,</span> <span class="n">processed_read</span> <span class="o">=</span> <span class="n">read_grouping</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span> <span class="n">fa</span><span class="p">)</span>
<span class="n">ref_exon</span><span class="p">,</span> <span class="n">ref_junction</span><span class="p">,</span> <span class="n">ref_single_exon_trans</span><span class="p">,</span> <span class="n">ref_mutple_exon_trans</span><span class="p">,</span> <span class="n">left_sj_set</span><span class="p">,</span> <span class="n">right_sj_set</span><span class="p">,</span> <span class="n">tss_dict</span> <span class="o">=</span> <span class="n">RefAnnotationExtraction</span><span class="p">(</span><span class="n">gtf</span><span class="p">)</span><span class="o">.</span><span class="n">annotation_sorting</span><span class="p">()</span>

<span class="c1"># polya_dict = polya_signal_import(&#39;/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058584.polya.res&#39;)</span>
<span class="c1"># polya_dict = PolyAFinder(processed_read, fa, &#39;/home/zjzace/software/SQANTI3-4.1/data/polyA_motifs/mouse_and_human.polyA_motif.txt&#39;).polya_estimation()</span>
<span class="c1"># collected_single_exon_read, collected_multi_exon_read, tss, rss = CoCoWrapper(16, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, tmp_dir=&#39;.&#39;, sj_correction_window=40, mis_intron_length = 150, corExcept_dis=4, polya_dict=polya_dict).result_collection()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="LafiteWrapper" class="doc_header"><code>class</code> <code>LafiteWrapper</code><a href="https://github.com/TF-Chan-Lab/LAFITE/tree/master/LAFITE/lafite_wrapper.py#L23" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>LafiteWrapper</code>(<strong><code>bam</code></strong>, <strong><code>bedtools</code></strong>, <strong><code>full_cleanup</code></strong>, <strong><code>gtf</code></strong>, <strong><code>genome</code></strong>, <strong><code>min_count_tss_tes</code></strong>, <strong><code>mis_intron_length</code></strong>, <strong><code>min_novel_trans_count</code></strong>, <strong><code>min_single_exon_coverage</code></strong>, <strong><code>min_single_exon_len</code></strong>, <strong><code>label</code></strong>, <strong><code>output</code></strong>, <strong><code>polya</code></strong>, <strong><code>polyA_motif_file</code></strong>, <strong><code>relative_abundance_threshold</code></strong>, <strong><code>sj_correction_window</code></strong>, <strong><code>short_sj_tab</code></strong>, <strong><code>thread</code></strong>, <strong><code>tss_cutoff</code></strong>, <strong><code>tss_peak</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">class</span> <span class="nc">LafiteWrapper</span><span class="p">:</span>
	<span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">bedtools</span><span class="p">,</span> <span class="n">full_cleanup</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">min_count_tss_tes</span><span class="p">,</span> <span class="n">mis_intron_length</span><span class="p">,</span> <span class="n">min_novel_trans_count</span><span class="p">,</span> <span class="n">min_single_exon_coverage</span><span class="p">,</span> <span class="n">min_single_exon_len</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">output</span><span class="p">,</span> <span class="n">polya</span><span class="p">,</span> <span class="n">polyA_motif_file</span><span class="p">,</span> <span class="n">relative_abundance_threshold</span><span class="p">,</span> <span class="n">sj_correction_window</span><span class="p">,</span> <span class="n">short_sj_tab</span><span class="p">,</span> <span class="n">thread</span><span class="p">,</span> <span class="n">tss_cutoff</span><span class="p">,</span> <span class="n">tss_peak</span><span class="p">):</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">bam</span> <span class="o">=</span> <span class="n">bam</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">bedtools</span> <span class="o">=</span> <span class="n">bedtools</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">full_cleanup</span> <span class="o">=</span> <span class="n">full_cleanup</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">gtf</span> <span class="o">=</span> <span class="n">gtf</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">min_count_tss_tes</span> <span class="o">=</span> <span class="n">min_count_tss_tes</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">mis_intron_length</span> <span class="o">=</span> <span class="n">mis_intron_length</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">min_novel_trans_count</span> <span class="o">=</span> <span class="n">min_novel_trans_count</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">min_single_exon_coverage</span> <span class="o">=</span> <span class="n">min_single_exon_coverage</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">min_single_exon_len</span> <span class="o">=</span> <span class="n">min_single_exon_len</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="n">label</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">output</span> <span class="o">=</span> <span class="n">output</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">polya</span> <span class="o">=</span> <span class="n">polya</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">polyA_motif_file</span> <span class="o">=</span> <span class="n">polyA_motif_file</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">relative_abundance_threshold</span> <span class="o">=</span> <span class="n">relative_abundance_threshold</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">sj_correction_window</span> <span class="o">=</span> <span class="n">sj_correction_window</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">short_sj_tab</span> <span class="o">=</span> <span class="n">short_sj_tab</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">thread</span> <span class="o">=</span> <span class="n">thread</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">tss_cutoff</span> <span class="o">=</span> <span class="n">tss_cutoff</span>
		<span class="bp">self</span><span class="o">.</span><span class="n">tss_peak</span> <span class="o">=</span> <span class="n">tss_peak</span>
		
	
	<span class="k">def</span> <span class="nf">revisit_parameter</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		revisit input parameters&quot;&quot;&quot;</span>

		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Input parameters:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Read alignment file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">bam</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Reference gene annotation: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gtf</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Reference genome annotation: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">polya</span><span class="p">:</span>
			<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Reads Polyadenylation events: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">polya</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">polyA_motif_file</span><span class="p">:</span>
			<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PolyA motif file for read Polyadenylation event estimation: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">polyA_motif_file</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Fatal: Please provide either polyA motif file or reads polyadenylation event result</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Edit distance to reference splicing site allowed for splicing correction: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">sj_correction_window</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Output assembly file: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

	<span class="k">def</span> <span class="nf">run_lafite</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
		<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">		LAFITE wrapper&quot;&quot;&quot;</span>

		<span class="bp">self</span><span class="o">.</span><span class="n">revisit_parameter</span><span class="p">()</span>
		<span class="c1"># create temp directory for log and intermediate files</span>
		<span class="k">try</span><span class="p">:</span>
			<span class="n">tmp_folder</span> <span class="o">=</span> <span class="n">temp_dir_creation</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">))</span>
			<span class="n">tmp_dir</span> <span class="o">=</span> <span class="n">tmp_folder</span><span class="o">.</span><span class="n">name</span>
		<span class="k">except</span><span class="p">:</span>
			<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Fatal: Please provide a valid path for output files</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

		<span class="c1"># reference gene annotation processing</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;: Preprocessing reference gene annotation</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">ref_exon</span><span class="p">,</span> <span class="n">ref_junction</span><span class="p">,</span> <span class="n">ref_single_exon_trans</span><span class="p">,</span> <span class="n">ref_mutple_exon_trans</span><span class="p">,</span> <span class="n">left_sj_set</span><span class="p">,</span> <span class="n">right_sj_set</span><span class="p">,</span> <span class="n">tss_dict</span> <span class="o">=</span> <span class="n">RefProcessWrapper</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gtf</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">thread</span><span class="p">)</span><span class="o">.</span><span class="n">result_collection</span><span class="p">()</span>

		<span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">short_sj_tab</span><span class="p">:</span>
			<span class="n">left_sj_set</span><span class="p">,</span> <span class="n">right_sj_set</span> <span class="o">=</span> <span class="n">short_reads_sj_import</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">short_sj_tab</span><span class="p">,</span> <span class="n">left_sj_set</span><span class="p">,</span> <span class="n">right_sj_set</span><span class="p">)</span>
		
		<span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_peak</span><span class="p">:</span>
			<span class="n">tss_dict</span> <span class="o">=</span> <span class="n">cage_tss_import</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tss_peak</span><span class="p">,</span> <span class="n">tss_dict</span><span class="p">)</span>

		<span class="c1"># processing alignment bam file</span>
		<span class="c1">## convert alignment bam file to bed12 format</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;: Preprocessing alignment file</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="k">try</span><span class="p">:</span> 
			<span class="n">bam2bed_cmd</span> <span class="o">=</span> <span class="n">bam2bed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bam</span><span class="p">,</span> <span class="n">tmp_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bedtools</span><span class="p">)</span>
		<span class="k">except</span><span class="p">:</span>
			<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Fatal: Please provide a valid path for bam file and bedtools</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">bam2bed_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
		<span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">returncode</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
			<span class="k">pass</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Fatal: Error in bam conversion</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

		<span class="c1">## read grouping according to chromosome and strand</span>
		<span class="n">outbed</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tmp_dir</span><span class="p">,</span> <span class="s1">&#39;bam.bed&#39;</span><span class="p">)</span>
		<span class="n">junction_dict</span><span class="p">,</span> <span class="n">processed_read</span> <span class="o">=</span> <span class="n">read_grouping</span><span class="p">(</span><span class="n">outbed</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>

		<span class="c1"># polyA info import</span>
		<span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">polya</span><span class="p">:</span>
			<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;: Loading polyA information</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
			<span class="n">polya_dict</span> <span class="o">=</span> <span class="n">polya_signal_import</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">polya</span><span class="p">)</span>
		<span class="k">else</span><span class="p">:</span>
			<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;: No reads Polyadenylation event provided, detecting from sequence</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
			<span class="n">polya_dict</span> <span class="o">=</span> <span class="n">PolyAFinder</span><span class="p">(</span><span class="n">processed_read</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">polyA_motif_file</span><span class="p">)</span><span class="o">.</span><span class="n">polya_estimation</span><span class="p">()</span>

		<span class="c1"># read correction and collapsing</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;: Collapssing corrected reads</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">collected_single_exon_read</span><span class="p">,</span> <span class="n">collected_multi_exon_read</span><span class="p">,</span> <span class="n">collected_rss</span><span class="p">,</span> <span class="n">collected_res</span> <span class="o">=</span> <span class="n">CoCoWrapper</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">processed_read</span><span class="p">,</span> <span class="n">ref_exon</span><span class="p">,</span> <span class="n">ref_junction</span><span class="p">,</span> <span class="n">ref_single_exon_trans</span><span class="p">,</span> <span class="n">ref_mutple_exon_trans</span><span class="p">,</span> <span class="n">left_sj_set</span><span class="p">,</span> <span class="n">right_sj_set</span><span class="p">,</span> <span class="n">junction_dict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sj_correction_window</span><span class="p">,</span> <span class="n">polya_dict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mis_intron_length</span><span class="p">,</span> <span class="n">tmp_dir</span><span class="p">)</span><span class="o">.</span><span class="n">result_collection</span><span class="p">()</span>

		<span class="c1"># identify putative TSS and TES for collapsed reads</span>
		<span class="n">processed_collected_multi_exon_read</span><span class="p">,</span> <span class="n">three_prime_exon</span> <span class="o">=</span> <span class="n">TailFinderWrapper</span><span class="p">(</span><span class="n">collected_multi_exon_read</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_count_tss_tes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">thread</span><span class="p">)</span><span class="o">.</span><span class="n">result_collection</span><span class="p">()</span>

		<span class="c1"># calculating the tss_cutoff and tes_cutoff:</span>
		<span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_cutoff</span><span class="p">:</span>
			<span class="bp">self</span><span class="o">.</span><span class="n">tss_cutoff</span> <span class="o">=</span> <span class="n">percentile</span><span class="p">(</span><span class="n">collected_rss</span><span class="p">,</span> <span class="mi">75</span><span class="p">)</span>
		<span class="n">tes_cutoff</span> <span class="o">=</span> <span class="n">percentile</span><span class="p">(</span><span class="n">collected_res</span><span class="p">,</span> <span class="mi">70</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;TSS cutoff: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">tss_cutoff</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;TES cutoff: </span><span class="si">{</span><span class="n">tes_cutoff</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

		<span class="c1"># identify high quality isoforms from collapsed reads</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;: Revisiting the collapsed reads to get high-concensus full-length isoforms</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
		<span class="n">collected_refined_isoforms</span> <span class="o">=</span> <span class="n">RefineWrapper</span><span class="p">(</span><span class="n">processed_collected_multi_exon_read</span><span class="p">,</span> <span class="n">collected_single_exon_read</span><span class="p">,</span> <span class="n">ref_mutple_exon_trans</span><span class="p">,</span> <span class="n">ref_single_exon_trans</span><span class="p">,</span> <span class="n">three_prime_exon</span><span class="p">,</span> <span class="n">tss_dict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_cutoff</span><span class="p">,</span> <span class="n">tes_cutoff</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_novel_trans_count</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_single_exon_coverage</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_single_exon_len</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">tmp_dir</span><span class="p">)</span><span class="o">.</span><span class="n">result_collection</span><span class="p">()</span>

		<span class="c1"># output refined isoforms</span>
		<span class="n">OutputAssembly</span><span class="p">(</span><span class="n">collected_refined_isoforms</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">relative_abundance_threshold</span><span class="p">)</span><span class="o">.</span><span class="n">write_out</span><span class="p">()</span>

		<span class="c1"># keep intermediate results</span>
		<span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">full_cleanup</span><span class="p">:</span>
			<span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">keep_tmp_file</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="n">tmp_dir</span><span class="p">))</span>
		
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;: All done!</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="main" class="doc_header"><code>main</code><a href="https://github.com/TF-Chan-Lab/LAFITE/tree/master/LAFITE/lafite_wrapper.py#L141" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>main</code>()</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>

	<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Low-abundance Aware Full-length Isoform clusTEr&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-b&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;bam&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the alignment file in bam format&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-B&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;bedtools&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;bedtools&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the executable bedtools&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;gtf&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the reference gene annotation in GTF format&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;genome&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the reference genome fasta&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;output&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the output file&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-n&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;min_count_tss_tes&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;minimum number of reads supporting a alternative TSS or TES, default: 3&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;mis_intron_length&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;length cutoff for correcting unexpected small intron, default: 150&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-c&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;min_novel_trans_count&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;minimum occurrences required for a isoform from novel loci, default: 3&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;min_single_exon_coverage&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;minimum read coverage required for a novel single-exon transcript, default: 4&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-l&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;min_single_exon_len&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;minimum length for single-exon transcript, default: 100&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-L&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;label&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;LAFT&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;name prefix for output transcripts, default: LAFT&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-p&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;polya&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the file contains read Polyadenylation event&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;polyA_motif_file&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the polya motif file&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;relative_abundance_threshold&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;minimum abundance of the predicted multi-exon transcripts as a fraction of the total transcript assembled at a given locus, default: 0.01&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-j&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;short_sj_tab&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the short read splice junction file&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-w&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;sj_correction_window&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;edit distance to reference splicing site for splicing correction, default: 40&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--no_full_cleanup&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;full_cleanup&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;keep all intermediate files&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-t&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;thread&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;number of the threads, default: 4&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-T&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;tss_peak&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the TSS peak file&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-d&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;tss_cutoff&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;minimum TSS distance for a transcript to be considered as a novel transcript&#39;</span><span class="p">)</span>
	<span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

	<span class="n">LafiteWrapper</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bam</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">bedtools</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">full_cleanup</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">gtf</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">min_count_tss_tes</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">mis_intron_length</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">min_novel_trans_count</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">min_single_exon_coverage</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">min_single_exon_len</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">polya</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">polyA_motif_file</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">relative_abundance_threshold</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">sj_correction_window</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">short_sj_tab</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">tss_cutoff</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">tss_peak</span><span class="p">)</span><span class="o">.</span><span class="n">run_lafite</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

