---

title: lafite_wrapper


keywords: fastai
sidebar: home_sidebar



nb_path: "10_wrapper.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 10_wrapper.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">LAFITE.reference_processing</span> <span class="kn">import</span> <span class="n">RefAnnotationExtraction</span>
<span class="kn">from</span> <span class="nn">LAFITE.preprocessing</span> <span class="kn">import</span> <span class="n">split_bed_line</span><span class="p">,</span> <span class="n">read_grouping</span><span class="p">,</span> <span class="n">polya_signal_import</span><span class="p">,</span> <span class="n">PolyAFinder</span>
<span class="kn">from</span> <span class="nn">LAFITE.utils</span> <span class="kn">import</span> <span class="n">loc_distance</span><span class="p">,</span> <span class="n">Vividict</span><span class="p">,</span> <span class="n">temp_dir_creation</span><span class="p">,</span> <span class="n">bam2bed</span>
<span class="kn">from</span> <span class="nn">LAFITE.read_collapsing</span> <span class="kn">import</span> <span class="n">CoCoWrapper</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span><span class="p">,</span> <span class="n">Counter</span><span class="p">,</span> <span class="n">OrderedDict</span>
<span class="kn">from</span> <span class="nn">interlap</span> <span class="kn">import</span> <span class="n">InterLap</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">time</span> <span class="kn">import</span> <span class="n">strftime</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">pysam</span><span class="o">,</span> <span class="nn">sys</span> <span class="o">,</span><span class="nn">os</span><span class="o">,</span> <span class="nn">subprocess</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="lafite_wrapper" class="doc_header"><code>lafite_wrapper</code><a href="https://github.com/zjzace/LAFITE/tree/master/LAFITE/lafite_wrapper.py#L22" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>lafite_wrapper</code>(<strong><code>bam</code></strong>, <strong><code>bedtools</code></strong>, <strong><code>gtf</code></strong>, <strong><code>genome</code></strong>, <strong><code>polya</code></strong>, <strong><code>polyA_motif_file</code></strong>, <strong><code>output</code></strong>, <strong><code>sj_correction_window</code></strong>=<em><code>40</code></em>, <strong><code>thread</code></strong>=<em><code>1</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">lafite_wrapper</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="n">bedtools</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">polya</span><span class="p">,</span> <span class="n">polyA_motif_file</span><span class="p">,</span> <span class="n">output</span><span class="p">,</span> <span class="n">sj_correction_window</span><span class="o">=</span> <span class="mi">40</span><span class="p">,</span> <span class="n">thread</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
	<span class="c1"># print out input parameters</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Input parameters:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;read alignment file: </span><span class="si">{</span><span class="n">bam</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Reference gene annotation: </span><span class="si">{</span><span class="n">gtf</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Reference genome annotation: </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Reads Polyadenylation events: </span><span class="si">{</span><span class="n">polya</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Output assembly file: </span><span class="si">{</span><span class="n">output</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Edit distance to reference splicing site allowed for splicing correction: </span><span class="si">{</span><span class="n">sj_correction_window</span><span class="si">}</span><span class="se">\n\n</span><span class="s1">&#39;</span><span class="p">)</span>

	<span class="c1"># create temp directory for log and intermediate files</span>
	<span class="k">try</span><span class="p">:</span>
		<span class="n">output_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output</span><span class="p">)</span>
		<span class="n">tmp_dir</span> <span class="o">=</span> <span class="n">temp_dir_creation</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
	<span class="k">except</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Fatal: please provide a valid path for output files</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

	<span class="c1"># reference gene annotation processing</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;: Preprocessing reference gene annotation</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
	<span class="n">ref_exon</span><span class="p">,</span> <span class="n">ref_junction</span><span class="p">,</span> <span class="n">ref_single_exon_trans</span><span class="p">,</span> <span class="n">ref_mutple_exon_trans</span><span class="p">,</span> <span class="n">left_sj_set</span><span class="p">,</span> <span class="n">right_sj_set</span><span class="p">,</span> <span class="n">tss_dict</span> <span class="o">=</span> <span class="n">RefAnnotationExtraction</span><span class="p">(</span><span class="n">gtf</span><span class="p">)</span><span class="o">.</span><span class="n">annotation_sorting</span><span class="p">()</span>

	
	<span class="c1"># processing alignment bam file</span>

	<span class="c1">## convert alignment bam file to bed12 format</span>
	<span class="k">try</span><span class="p">:</span> 
		<span class="n">bam2bed_cmd</span> <span class="o">=</span> <span class="n">bam2bed</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="n">tmp_dir</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">bedtools</span><span class="p">)</span>
	<span class="k">except</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Fatal: please provide a valid path for bam file, samtools and bedtools</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
	
	<span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">bam2bed_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">returncode</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
		<span class="k">pass</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Fatal: error in bam conversion&quot;</span><span class="p">)</span>
	
	<span class="c1">## read grouping according to chromosome and strand</span>
	<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;: Preprocessing read alignment file</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
	<span class="n">outbed</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s1">&#39;bam.bed&#39;</span><span class="p">)</span>
	<span class="n">junction_dict</span><span class="p">,</span> <span class="n">processed_read</span> <span class="o">=</span> <span class="n">read_grouping</span><span class="p">(</span><span class="n">outbed</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span>

	<span class="c1"># polyA info import</span>
	<span class="k">if</span> <span class="n">polya</span><span class="p">:</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;: Loading polyA information</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
		<span class="n">polya_dict</span> <span class="o">=</span> <span class="n">polya_signal_import</span><span class="p">(</span><span class="n">polya</span><span class="p">)</span>
	<span class="k">else</span><span class="p">:</span>
		<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;: No reads Polyadenylation event provided, detecting from sequencing</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
		<span class="n">polya_dict</span> <span class="o">=</span> <span class="n">PolyAFinder</span><span class="p">(</span><span class="n">processed_read</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">polyA_motif_file</span><span class="p">)</span><span class="o">.</span><span class="n">polya_estimation</span><span class="p">()</span>
	
	<span class="c1"># read correction and collapsing</span>
	<span class="n">collected_single_exon_read</span><span class="p">,</span> <span class="n">collected_multi_exon_read</span> <span class="o">=</span> <span class="n">CoCoWrapper</span><span class="p">(</span><span class="n">thread</span><span class="p">,</span> <span class="n">processed_read</span><span class="p">,</span> <span class="n">ref_exon</span><span class="p">,</span> <span class="n">ref_junction</span><span class="p">,</span> <span class="n">ref_single_exon_trans</span><span class="p">,</span> <span class="n">ref_mutple_exon_trans</span><span class="p">,</span> <span class="n">left_sj_set</span><span class="p">,</span> <span class="n">right_sj_set</span><span class="p">,</span> <span class="n">junction_dict</span><span class="p">,</span> <span class="n">sj_correction_window</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">tmp_dir</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">mis_intron_length</span> <span class="o">=</span> <span class="mi">150</span><span class="p">,</span> <span class="n">corExcept_dis</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">polya_dict</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">result_collection</span><span class="p">()</span>


<span class="k">if</span> <span class="vm">__name__</span><span class="o">==</span><span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
	<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Low-abundance Aware Full-length Isoform clusTEr&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-a&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;bam&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the alignment file in bam format&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-b&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;bedtools&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the executable bedtools&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;gtf&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the reference gene annotation in GTF format&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;genome&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the reference genome fasta&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;output&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the output file&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-t&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;thread&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;number of the threads&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-p&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;polya&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the file contains read Polyadenylation event&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;polyA_motif_file&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to the polya motif file&#39;</span><span class="p">)</span>
	<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--sj_correction_window&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;sj_correction_window&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;edit distance to reference splicing site for splicing correction&#39;</span><span class="p">)</span>

	<span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

	<span class="n">lafite_wrapper</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bam</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">bedtools</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">gtf</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">polya</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">polyA_motif_file</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">sj_correction_window</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">thread</span><span class="p">)</span>

	
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

